### PR TITLE
Stop synchronisation when behind a captive portal

### PR DESCRIPTION
**Pull request by erikjv**
Fixes: #11533

### CODE REVIEW

**Comment by erikjv** [comment]
The `SpacesManager` still needs to be hooked up. So maybe the `Utility` code needs to move to libsync.

File: src/gui/guiutility.cpp [hunk 1]
```diff
@@ -100,6 +100,32 @@ QString Utility::vfsFreeSpaceActionText()
     return QCoreApplication::translate("utility", "Free up local space");
 }
 
+static bool captivePortalForcedOn = false;
```
Comment (TheOneRing) [inline src/gui/guiutility.cpp:103]: Can we add that somehow to our nam?

File: src/gui/scheduling/syncscheduler.cpp [hunk 2]
```diff
@@ -155,6 +155,11 @@ void SyncScheduler::startNext()
         return;
     }
 
+    if (Utility::internetThroughCaptivePortal()) {
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/scheduling/syncscheduler.cpp:158]: done

File: src/gui/accountstate.cpp [hunk 3]
```diff
@@ -171,6 +171,22 @@ AccountState::AccountState(AccountPtr account)
         ownCloudGui::raise();
         msgBox->open();
     });
+
+
+    connect(_account->accessManager(), &AccessManager::isBehindCaptivePortalChanged, this, [this](bool onoff) {
+        if (onoff) {
+            _queueGuard.block();
+        } else {
+            // TODO: empty queue?
```
Comment (TheOneRing) [inline src/gui/accountstate.cpp:180]: I'm not sure. It might depend on how long we were blocked, 10s or 20h...

File: src/gui/accountstate.cpp [hunk 4]
```diff
@@ -171,6 +171,22 @@ AccountState::AccountState(AccountPtr account)
...
+            _queueGuard.unblock();
+        }
+    });
+    if (_account->accessManager()->isBehindCaptivePortal()) {
+        _queueGuard.block();
+    } else {
+        // TODO: empty queue?
```
Comment (TheOneRing) [inline src/gui/accountstate.cpp:187]: I don't think this unblock has any use?

File: src/gui/accountstate.cpp [hunk 5]
```diff
@@ -247,6 +263,7 @@ void AccountState::setState(State state)
         } else if (_state == Connected && Utility::internetConnectionIsMetered() && ConfigFile().pauseSyncWhenMetered()) {
             _state = PausedDueToMetered;
         }
+        // Do we need an extra state PausedDueToCaptivePortal?
```
Comment (TheOneRing) [inline src/gui/accountstate.cpp:266]: For now we could reuse "Connecting"? In the end its only to display a different string? In which case we could make the "connecting" message conditional on isBehindCaptivePortal

File: src/gui/main.cpp [hunk 6]
```diff
@@ -85,7 +85,7 @@ struct CommandLineOptions
     bool logFlush = false;
     bool logDebug = false;
 
-    bool debugMode = false;
```
Comment (TheOneRing) [inline src/gui/main.cpp:88]: Nope.
Just start the client with "--debug"

File: src/gui/owncloudgui.cpp [hunk 7]
```diff
@@ -641,17 +641,31 @@ void ownCloudGui::updateContextMenu()
...
-        crashMenu->addAction(QStringLiteral("Crash now - OC_ENFORCE()"), _app, [] { OC_ENFORCE(false); });
-        crashMenu->addAction(QStringLiteral("Crash now - qFatal"), _app, [] { qFatal("la Qt fatale"); });
-        crashMenu->addAction(QStringLiteral("Restart now"), _app, [] { RestartManager::requestRestart(); });
+        debugMenu->addAction(QStringLiteral("Crash if asserts enabled - Q_ASSERT"), _app, [] { Q_ASSERT(false); });
+        debugMenu->addAction(QStringLiteral("Crash now - Utility::crash()"), _app, [] { Utility::crash(); });
+        debugMenu->addAction(QStringLiteral("Crash now - OC_ENFORCE()"), _app, [] { OC_ENFORCE(false); });
+        debugMenu->addAction(QStringLiteral("Crash now - qFatal"), _app, [] { qFatal("la Qt fatale"); });
+        debugMenu->addAction(QStringLiteral("Restart now"), _app, [] { RestartManager::requestRestart(); });
+        debugMenu->addSeparator();
+        auto captivePortalCheckbox = debugMenu->addAction(QStringLiteral("Behind Captive Portal"));
```
Comment (TheOneRing) [inline src/gui/owncloudgui.cpp:656]: For a real life scenario we also need to see the tls issue, did you test with one of the ocis examples?

Comment (TheOneRing) [reply src/gui/owncloudgui.cpp:656]: For the style, maybe use a direct connection like for the other actions?

File: src/libsync/accessmanager.cpp [hunk 8]
```diff
@@ -35,6 +36,7 @@ Q_LOGGING_CATEGORY(lcAccessManager, "sync.accessmanager", QtInfoMsg)
 
 AccessManager::AccessManager(QObject *parent)
     : QNetworkAccessManager(parent)
+    , _forcedCaptivePortal(false)
```
Comment (TheOneRing) [inline src/libsync/accessmanager.cpp:39]: Please initialize in the header

**Review by TheOneRing — CHANGES_REQUESTED** [review]
For a manual testing action, it has quite a heavy impact on the client...

**Comment by TheOneRing** [comment]
Brainstorming: We have multiple network information things to test, could it make sense to implement a custom one that wraps the native one and allows overriding values?

**Comment by TheOneRing** [comment]
Please rebase, once done I can start a review.

File: src/gui/accountstate.cpp [hunk 9]
```diff
@@ -113,42 +113,45 @@ AccountState::AccountState(AccountPtr account)
...
-        });
-    }
-#endif
+        }
+    });
+
+    connect(NetworkInformation::instance(), &NetworkInformation::isBehindCaptivePortalChanged, this, [this](bool) {
+        // A direct connect is not possible, because then the state parameter of `isBehindCaptivePortalChanged`
+        // would become the `verifyServerState` argument to `checkConnectivity`.
+        QTimer::singleShot(0, this, [this] { checkConnectivity(false); });
```
Comment (TheOneRing) [inline src/gui/accountstate.cpp:152]: Does it make sense to run the check if it changed to true?

Comment (erikjv) [reply src/gui/accountstate.cpp:152]: Yes, it sets the state to `Connecting` and triggers the signals so the UI shows paused and captive portal.

File: src/gui/accountstate.cpp (hunk 3 above)
Comment (erikjv) [reply src/gui/accountstate.cpp:181]: I'm still not sure what to do with this

File: src/gui/accountstate.cpp (hunk 9 above)
Comment (TheOneRing) [reply src/gui/accountstate.cpp:152]: Could you add that as a comment? I see the possibility that someone might "Optimize" that call.

File: test/testutils/testutils.cpp [hunk 10]
```diff
@@ -27,6 +28,8 @@ namespace OCC {
 namespace TestUtils {
     TestUtilsPrivate::AccountStateRaii createDummyAccount()
     {
+        // ensure we have an instance of NetworkInformation
+        NetworkInformation::initialize();
```
Comment (TheOneRing) [inline test/testutils/testutils.cpp:32]: Should we move that to https://github.com/owncloud/client/blob/master/test/testutils/testutilsloader.cpp#L9

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: src/gui/networkinformation.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+#include <QLoggingCategory>
+
+using namespace OCC;
+
+Q_LOGGING_CATEGORY(lcNetInfo, "gui.netinfo", QtInfoMsg)
+
+NetworkInformation *NetworkInformation::_instance = nullptr;
+
+static void loadQNetworkInformationBackend()
```
Comment (TheOneRing) [inline src/gui/networkinformation.cpp:25]: anonymous namespace please

File: src/gui/networkinformation.cpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    qCDebug(lcNetInfo) << "Supported features:" << QNetworkInformation::instance()->supportedFeatures();
+    qCDebug(lcNetInfo) << "Available backends:" << QNetworkInformation::availableBackends().join(QStringLiteral(", "));
+    if (auto qni = QNetworkInformation::instance()) {
+        QObject::connect(qni, &QNetworkInformation::reachabilityChanged,
+            [](QNetworkInformation::Reachability reachability) { qCInfo(lcNetInfo) << "Connection Status changed to:" << reachability; });
+    }
+}
+
+void NetworkInformation::initialize()
+{
```
Comment (TheOneRing) [inline src/gui/networkinformation.cpp:48]: Do we need the explicit initialize call or could it make sense to merge it with  the instance call?

File: src/gui/accountstate.cpp (hunk 3 above)
Comment (TheOneRing) [reply src/gui/accountstate.cpp:181]: Lets clear it.

File: src/gui/accountstate.cpp (hunk 4 above)
Comment (TheOneRing) [reply src/gui/accountstate.cpp:188]: How can we be blocked during the construction of the class?

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: src/gui/accountstate.cpp [hunk 13]
```diff
@@ -113,42 +113,57 @@ AccountState::AccountState(AccountPtr account)
...
             }
-        });
+        }
+    });
+
+    connect(NetworkInformation::instance(), &NetworkInformation::isBehindCaptivePortalChanged, this, [this](bool onoff) {
+        if (onoff) {
+            _queueGuard.block();
+        } else {
+            _account->jobQueue()->clear();
```
Comment (TheOneRing) [inline src/gui/accountstate.cpp:153]: Can you add a comment on why we decided to clear the queue?

**Review by TheOneRing — DISMISSED** [review]

**Comment by erikjv** [comment]
See the screenshot in https://github.com/owncloud/client/issues/11533#issuecomment-2288816029

**Review by TheOneRing — APPROVED** [review]

**Pull request merged**